File: cs-499 project java/Appointment/Appointment.java
/*
 * Name: Justin Fifarek
 * Course: SNHU CS-320
 * Date: 12/01/2023
 * 
 * Module 5-1 Milestone - Appointment class
 */

package Appointment;

import java.util.concurrent.atomic.AtomicLong;
import java.util.Calendar;
import java.util.Date;

public class Appointment {
	// Private variables for appointment information
	private final String appointmentID;
	private Date appointmentDate;
	private String appointmentDesc;
	private static AtomicLong idGenerator = new AtomicLong();
	
	@SuppressWarnings("deprecation")
	public Appointment(Date appointmentDate, String appointmentDesc) {
		// appointmentID is generated when constructor is called
		this.appointmentID = String.valueOf(idGenerator.getAndIncrement());
		
		// appointmentDate cannot be before the current date
		if (appointmentDate == null) {
			this.appointmentDate = new Date(2025, Calendar.JANUARY, 1);
		}
		else if (appointmentDate.before(new Date())) {
			throw new IllegalArgumentException("Invalid date. Date cannot be in the past.");
		}
		else {
			this.appointmentDate = appointmentDate;
		}
		
		// appointmentDesc cannot be more than 50 characters
		// and cannot be null
		if (appointmentDesc == null || appointmentDesc.isEmpty()) {
			this.appointmentDesc = appointmentDesc;
		}
		else if (appointmentDesc.length() > 50) {
			this.appointmentDesc = appointmentDesc.substring(0, 50);
		}
		else {
			this.appointmentDesc = appointmentDesc;
		}
	}
	
	// Getters
	public String getAppointmentID() {
		return this.appointmentID;
	}
	public Date getAppointmentDate() {
		return this.appointmentDate;
	}
	public String getAppointmentDesc() {
		return this.appointmentDesc;
	}
	
	// Setters
	// Set up the same as in constructor
	@SuppressWarnings("deprecation")
	public void setAppointmentDate(Date appointmentDate) {
		if (appointmentDate == null) {
			this.appointmentDate = new Date(2024, Calendar.JANUARY, 1);
		}
		else if (appointmentDate.before(new Date())) {
			throw new IllegalArgumentException("Invalid date. Date cannot be in the past.");
		}
		else {
			this.appointmentDate = appointmentDate;
		}
	}
	
	public void setAppointmentDesc(String appointmentDesc) {
		if (appointmentDesc == null || appointmentDesc.isEmpty()) {
			this.appointmentDesc = appointmentDesc;
		}
		else if (appointmentDesc.length() > 50) {
			this.appointmentDesc = appointmentDesc.substring(0, 50);
		}
		else {
			this.appointmentDesc = appointmentDesc;
		}
	}
}

File: cs-499 project c++/Appointment/Appointment.cpp
/*
    Name: Justin Fifarek
    Project: Appointment class
*/
#include <iostream>
#include <string>
#include <atomic>
#include <ctime>

class Appointment {
private:
    // Private variables for appointment information
    std::string appointmentID;
    std::time_t appointmentDate;
    std::string appointmentDesc;
    static std::atomic<long> idGenerator;

public:
    Appointment(std::time_t appointmentDate, std::string appointmentDesc) {
        // appointmentID is generated when constructor is called
        this->appointmentID = std::to_string(idGenerator++);

        // appointmentDate cannot be before the current date
        std::time_t now = std::time(nullptr);
        if (appointmentDate < now) {
            throw std::invalid_argument("Invalid date. Date cannot be in the past.");
        }
        this->appointmentDate = appointmentDate;

        // appointmentDesc cannot be more than 50 characters
        // and cannot be null
        if (appointmentDesc.empty()) {
            this->appointmentDesc = appointmentDesc;
        }
        else if (appointmentDesc.length() > 50) {
            this->appointmentDesc = appointmentDesc.substr(0, 50);
        }
        else {
            this->appointmentDesc = appointmentDesc;
        }
    }

    // Getters
    std::string getAppointmentID() {
        return appointmentID;
    }

    std::time_t getAppointmentDate() {
        return appointmentDate;
    }

    std::string getAppointmentDesc() {
        return appointmentDesc;
    }

    // Setters
    // Set up the same as in constructor
    void setAppointmentDate(std::time_t appointmentDate) {
        std::time_t now = std::time(nullptr);
        if (appointmentDate < now) {
            throw std::invalid_argument("Invalid date. Date cannot be in the past.");
        }
        this->appointmentDate = appointmentDate;
    }

    void setAppointmentDesc(const std::string& appointmentDesc) {
        if (appointmentDesc.empty()) {
            this->appointmentDesc = appointmentDesc;
        }
        else if (appointmentDesc.length() > 50) {
            this->appointmentDesc = appointmentDesc.substr(0, 50);
        }
        else {
            this->appointmentDesc = appointmentDesc;
        }
    }
};
